src/container/container_iv.c

```diff
@@ -0,0 +1,415 @@
+/**
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * container IV cache
+ */
+#define D_LOGFAC	DD_FAC(container)
+
+#include <daos_srv/container.h>
+#include "srv_internal.h"
+#include <daos_srv/iv.h>
+#include <daos/btree_class.h>
+#include <daos/btree.h>
+#include <daos/dtx.h>
+
+/* XXX Temporary limit for IV */
+#define MAX_SNAP_CNT	20
+/* Container IV structure */
+struct cont_iv_entry {
+	uuid_t	cont_uuid;
+	int	snap_size;
+	uint64_t snaps[0];
+};
+
+struct cont_iv_key {
+	uuid_t	cont_uuid;
+};
+
+static struct cont_iv_key *
+key2priv(struct ds_iv_key *iv_key)
+{
+	return (struct cont_iv_key *)iv_key->key_buf;
+}
+
+static uint32_t
+cont_iv_ent_size(int nr)
+{
+	return sizeof(struct cont_iv_entry) + nr * sizeof(uint64_t);
+}
+
+static int
+cont_iv_value_alloc_internal(d_sg_list_t *sgl)
+{
+	struct cont_iv_entry	*entry;
+	int			entry_size;
+	int			rc;
+
+	rc = daos_sgl_init(sgl, 1);
+	if (rc)
+		return rc;
+
+	/* FIXME: allocate entry by the real size */
+	entry_size = cont_iv_ent_size(MAX_SNAP_CNT);
+	D_ALLOC(entry, entry_size);
```

Comment:
sorry, I did not get where to free it, could you please remind me? thanks
it seems to me need do the free in cont_iv_ent_put(), but I'm not quite understand the IV code.

Reply:
oh, cont_iv_entry will be destroyed in cont_iv_ent_destroy(). see ivc_on_put()->iv_entry_free()->cont_iv_ent_destroy()